llama.cpp

```diff
@@ -3075,118 +3170,130 @@ struct llama_model_loader {
         }
     }
 
-    size_t file_offset(const char * name) const {
-        const int idx = gguf_find_tensor(ctx_gguf, name);
-
-        if (idx < 0) {
-            throw std::runtime_error(format("%s: tensor '%s' not found in the file", __func__, name));
-        }
-
-        return gguf_get_data_offset(ctx_gguf) + gguf_get_tensor_offset(ctx_gguf, idx);
-    }
-
-    void init_mapping(bool prefetch = true, llama_mlock * lmlock = nullptr) {
-        // prefetch the whole file - all the data is needed anyway
+    void init_mappings(bool prefetch = true, std::vector<std::unique_ptr<llama_mlock>> * mlock_mmaps = nullptr) {
         if (use_mmap) {
-            mapping.reset(new llama_mmap(&file, prefetch ? -1 : 0, ggml_is_numa()));
+            for (const auto & file : files) {
+                auto * mapping = new llama_mmap(file.get(), prefetch ? -1 : 0, ggml_is_numa());
+                mmaps_used.emplace_back(std::make_pair(mapping->size, 0));
+                mappings.emplace_back(std::unique_ptr<llama_mmap>(mapping));
```

Comment:
If `mmaps_used.emplace_back` throws, it will leak `mappings`.

Reply:
I have reserved enough cap before, please confirm this approach is OK

Reply:
Also as I understood, even with an exception, previous `mapping` will be deleted with the `llama_model_loader` deletion thanks to `std::unique_ptr` in the `mappinps` ?

Reply:
`mapping` is not in a `unique_ptr` in the second line, so it wouldn't be freed if it throws. A simple solution would be to put `mapping` in a `unique_ptr` from the moment it is allocated:
```
                std::unique_ptr<llama_mappa> mapping(new llama_mmap(file.get(), prefetch ? -1 : 0, ggml_is_numa()));
                mmaps_used.emplace_back(std::make_pair(mapping->size, 0));
                mappings.emplace_back(std::move(mapping));
```